### PR TITLE
Nihon Kohden EEG file reader WIP

### PR DESCRIPTION
**Pull request by fraimondo**
This is a reader module for Nihon Kohden native EEG file format.

It is a WIP, so for now it does not read events, no metadata and the channel names are the standard ones (it can be renamed through a file).

I still need to add a test. 

I leave it here so we can start doing the collaborative stuff and if anyone has any interest, they can chip in.

### CODE REVIEW

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 4b922d80ad027ca3147f656e50eb68b6c18c4a3e into ce50c2b20ac1197d79e68665900c9f6646505b13 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-21f4acbe6b338f8054abbe6831e15072297ba97f)

**new alerts:**

* 1 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
@fraimondo can you get some tiny file in `mne-testing-data`? I don't have any Nihon Koden data, but once that testing file is there, I (and others) can help with this.

**Comment by fraimondo** [comment]
I need to check who has the system that can provide me with a testing dataset. I’ll come back soon.

**Comment by fraimondo** [comment]
Just created a PR in mne-testing-data.

**Comment by larsoner** [comment]
@fraimondo any time to come back to this in the next couple of weeks to make it into 0.21? You now have your testing file in `mne-testing-data`.

If not, okay for me to push some commits to get tests working?

**Comment by fraimondo** [comment]
Just updated the code. This works. Can be improved by reading the annotations and more metadata.

**Comment by fraimondo** [comment]
There's an issue with the channel names. Hold the PR.

**Comment by larsoner** [comment]
> Just updated the code. This works. Can be improved by reading the annotations and more metadata.

Awesome, I was just about to look into this to see if we could get into usable shape in the next week to make it into 0.21. Let me know if you need help

**Comment by fraimondo** [comment]
Can't figure out this one. I'm getting `None` as the `cals` on the `_read_segment_file` function. I guess it's related to this commit: https://github.com/mne-tools/mne-python/commit/06f232cd75c203765b7165a8d94df1a52d5c7cea

So now I use the `_mult_cal_one` function. But it still fails. Apparently the length of the `idx` parameter does not match `data`.

It gives me an error that it can't broadcast the data in `_mult_cal_one`. So here's the issue:
```
> /Users/fraimondo/dev/mne-python/mne/io/nihon/nihon.py(394)_read_segment_file()
    392         import pdb; pdb.set_trace()
    393         # block_data = ((block_data[idx, :] * cals + offsets) * gains)
--> 394         _mult_cal_one(data, block_data, idx, cals, mult)
    395         data[:, :] += offsets
    396         data[:, :] *= gains

ipdb> data.shape                                                                                                                        
(1, 5800)
ipdb> idx.shape                                                                                                                         
(24,)
```

any clue?

**Comment by larsoner** [comment]
I recently made some changes to this function. Okay for me to look and push a commit tomorrow?

**Comment by fraimondo** [comment]
Feel free. PR is ready to be merged one this is solved.

The only solution I have in mind is to set cals=1 and then do the `(*scaling + offset) * gains` before calling `_mult_cal_one`.

**Comment by larsoner** [comment]
Setting `cals=1` should be fine. The `edf` reader was updated to behave this way.

Basically the second arg of `_mult_cal_one` (`one`) should always have shape `(n_data_channels, n_times)` (and `n_times` can be shorter if you use it in chunks / with time-based slicing) then the `_mult_cal_one` will subindex the `idx` from `n_data_channels` that it needs. It's a bit wacky but it should work. The `idx` is really there just to tell you which rows of the `one` you actually need to fill / will be used.

**Comment by fraimondo** [comment]
Yes, I got that.

But when a projector is applied while picking channels (as in the `_test_raw_reader` function), the `mult` parameter has shape (n_proj, n_channels). So the way I was offsetting and multiplying by units does not work, since `data` will have shape (n_proj, n_times).

I’ll do the cal=1 trick.

**Comment by larsoner** [comment]
> But when a projector is applied while picking channels (as in the `_test_raw_reader` function), the `mult` parameter has shape (n_proj, n_channels). So the way I was offsetting and multiplying by units does not work, since `data` will have shape (n_proj, n_times).

You should not need to worry about the shape of `mult`. If you make sure `one` has shape `(n_channels, n_times)`, internally `_mult_cal_one` will subselect the correct channels after projection (this actually happens just by `mult @ one`, more or less). So your job as the reader-writer is to make sure `one` is of shape `(n_channels, n_times)` with the `idx` indices populated. Then everything should work. You shouldn't ever use `cals` yourself, and you should make sure that you know that they will be multiplied into the data values. So if your pattern is `one = (one * a + b) * c` as above, you can either set `cals` to `1.` for all channels in `info['ch'][ii]['cal']`, or you can set them to `c` in the expression above and only take care of the `one = (one * a + b)` part of the expression (and let `_mult_cal_one` do the `cals` mult for you). What you don't want is it happening in both places.

**Comment by agramfort** [comment]
CIs complain @fraimondo 

maybe rebase on master to fix it?

**Comment by fraimondo** [comment]
Codecov is not happy with the coverage (80%). It's just a few lines testing for proper headers in the files / right values in the right places. I don't really know how to test those lines.

The other two are unrelated to this PR:

```
================================== FAILURES ===================================
_________ test_make_inverse_operator_diag[testing_data-testing_data] __________
mne\minimum_norm\tests\test_inverse.py:749: in test_make_inverse_operator_diag
    _compare_inverses_approx(inverse_operator_diag, inv_op, evoked,
mne\minimum_norm\tests\test_inverse.py:183: in _compare_inverses_approx
    assert corr > ctol
E   assert 0.7766994175372354 > 0.99
```

**Comment by larsoner** [comment]
@fraimondo let me know when you're done and I'll see if I can simplify the `_mult_cal_one` stuff. I also see that `info['chs'][ii]['unit']` is being used in a way that it shouldn't -- I'll add a check for that and fix it in this PR

**Comment by fraimondo** [comment]
I'm fine now. I think I got enough codecov. At least I don't see a failed test.

File: mne/io/nihon/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+"""Nihon Kohden module for conversion to FIF."""
+
+# Author: Fede Raimondo <federaimondo@gmail.com>
+#
+# License: BSD (3-clause)
+
+from .nihon import read_raw_nihon
```
Comment (agramfort) [inline mne/io/nihon/__init__.py:7]: missing empty line at the end of file

File: mne/io/nihon/nihon.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+    'T3', 'T4', 'T5', 'T6', 'FZ', 'CZ', 'PZ', 'E', 'PG1', 'PG2', 'A1', 'A2',
+    'T1', 'T2'
+]
+_default_chan_labels += ['X{}'.format(i) for i in range(1, 12)]
+_default_chan_labels += ['NA{}'.format(i) for i in range(1, 6)]
+_default_chan_labels += ['DC{:02}'.format(i) for i in range(1, 33)]
+_default_chan_labels += ['BN1', 'BN2', 'Mark1', 'Mark2']
+_default_chan_labels += ['NA{}'.format(i) for i in range(6, 28)]
+_default_chan_labels += ['X12/BP1', 'X13/BP2', 'X14/BP3', 'X15/BP4']
+_default_chan_labels += ['X{}'.format(i) for i in range(16, 166)]
```
Comment (agramfort) [inline mne/io/nihon/nihon.py:97]: you could use fstrings now rather than format

File: mne/io/nihon/nihon.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+        raise NotImplementedError('I dont know how to read more than one '
+                                  'control block for this type of file :(')
+    if header['controlblocks'][0]['n_datablocks'] != 1:
+        raise NotImplementedError('I dont know how to read more than one '
+                                  'data block for this type of file :(')
+
+    return header
+
+
+def _read_nihon_events(fname, orig_time):
```
Comment (agramfort) [inline mne/io/nihon/nihon.py:221]:
```suggestion
def _read_nihon_annotations(fname, orig_time):
```

**Comment by larsoner** [comment]
@fraimondo pushed a commit to simplify. Feel free to address @agramfort's remaining comment then hopefully we're good to go!

File: mne/io/nihon/nihon.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+                           t_onset.second)
+                all_onsets.append(t_onset)
+                all_descriptions.append(t_desc)
+
+        annotations = Annotations(all_onsets, 0.0, all_descriptions, orig_time)
+    return annotations
+
+
+def _map_ch_to_type(ch_name):
+    ch_type_pattern = OrderedDict(pair for pair in [
```
Comment (larsoner) [inline mne/io/nihon/nihon.py:260]:
```suggestion
    ch_type_pattern = OrderedDict([
```

**Comment by agramfort** [comment]
Thx @fraimondo

**Comment by ftadel** [comment]
Hello,

A colleague posted a NK .EEG file I don't know how to read:
- https://neuroimage.usc.edu/forums/t/error-message-when-reviewing-raw-file-for-the-edf-files-generated-by-nihon-kohden/23798/11
- https://drive.google.com/file/d/1tGZGev8YZ6YooSZyLQfwKAbUSxV3ZL2u/view?usp=sharing

In case you have interesting sources of information about the NK file formats...

Thanks!
François

**Comment by agramfort** [comment]
MNE Python complains too:

```
In [3]: import mne
   ...:
   ...: raw = mne.io.read_raw_nihon('resting.eeg')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-c17d6c9c0e8b> in <module>
      1 import mne
      2
----> 3 raw = mne.io.read_raw_nihon('resting.eeg')

~/work/src/mne-python/mne/io/nihon/nihon.py in read_raw_nihon(fname, preload, verbose)
     44     mne.io.Raw : Documentation of attribute and methods.
     45     """
---> 46     return RawNihon(fname, preload, verbose)
     47
     48

<decorator-gen-224> in __init__(self, fname, preload, verbose)

~/work/src/mne-python/mne/io/nihon/nihon.py in __init__(self, fname, preload, verbose)
    313         logger.info('Loading %s' % data_name)
    314
--> 315         header = _read_nihon_header(fname)
    316         metadata = _read_nihon_metadata(fname)
    317

~/work/src/mne-python/mne/io/nihon/nihon.py in _read_nihon_header(fname)
    134         version = np.fromfile(fid, '|S16', 1).astype('U16')[0]
    135         if version not in _valid_headers:
--> 136             raise ValueError(
    137                 'Not a valid Nihon Kohden EEG file ({})'.format(version))
    138

ValueError: Not a valid Nihon Kohden EEG file (BFA NKC04/01/16)
```

**Pull request merged**